File: TerraObserver/TO.Domains/Functions/Units/HexUnitPathPool.fs
namespace TO.Domains.Functions.Units

open TO.Domains.Types.HexSpheres
open TO.Domains.Types.Units

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-07-11 09:47:11
module HexUnitPathPoolCommand =
    let private fetchPath (pool: IHexUnitPathPool) =
        let path =
            pool.Paths
            |> Seq.tryFind (fun p -> not p.Working)
            |> Option.defaultWith (fun () ->
                let p = pool.InstantiatePath()
                pool.Paths.Add p
                p)

        path.Working <- true
        path

    let newUnitPathTask (env: #IHexUnitPathPoolQuery) : NewUnitPathTask =
        fun (pathTiles: TileId array) ->
            let pathOpt = env.HexUnitPathPoolOpt |> Option.map fetchPath
            pathOpt |> Option.iter (HexUnitPathCommand.unitPathTaskStart env pathTiles)
            pathOpt

File: TerraObserver/Scenes/Units/Views/HexUnitPathPool.cs
using System.Collections.Generic;
using Godot;
using TO.Domains.Types.Units;

namespace TerraObserver.Scenes.Units.Views;

/// Copyright (C) 2025 Zhu Xiaohe(aka ZeromaXHe)
/// Author: Zhu XH (ZeromaXHe)
/// Date: 2025-07-11 09:34:10
public partial class HexUnitPathPool : Node3D, IHexUnitPathPool
{
    #region Export

    [Export] public PackedScene? PathScene { get; set; }

    public IHexUnitPath InstantiatePath()
    {
        var path = PathScene!.Instantiate<HexUnitPath>();
        AddChild(path);
        return path;
    }

    #endregion

    #region 普通属性

    public List<IHexUnitPath> Paths { get; } = [];

    #endregion
}